integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala

```diff
@@ -268,6 +309,151 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
+              s"${CarbonCommonConstants.INDEX_HANDLER}.$e.sourcecolumns property is not specified.")
+        } else if (sourceColumnsOption.get.isEmpty) {
+          throw new MalformedCarbonCommandException(
+            s"Carbon ${CarbonCommonConstants.INDEX_HANDLER} property is invalid. " +
+              s"${CarbonCommonConstants.INDEX_HANDLER}.$e.sourcecolumns property cannot be empty.")
+        }
+
+        /* Validate source columns */
+        val sources = sourceColumnsOption.get.split(",")
+        if (sources.distinct.length != sources.size) {
+          throw new MalformedCarbonCommandException(
+            s"Carbon ${CarbonCommonConstants.INDEX_HANDLER} property is invalid. " +
+              s"${CarbonCommonConstants.INDEX_HANDLER}.$e.sourcecolumns property " +
+              s"have duplicate columns.")
+        }
+
+        val sourceTypes = StringBuilder.newBuilder
+        sources.foreach { column =>
+          tableFields.find(_.column.equalsIgnoreCase(column)) match {
+            case Some(field) => sourceTypes.append(field.dataType.get).append(",")
+            case None =>
+              throw new MalformedCarbonCommandException(
+                s"Carbon ${CarbonCommonConstants.INDEX_HANDLER} property is invalid. " +
+                  s"Source column : $column in property " +
+                  s"${CarbonCommonConstants.INDEX_HANDLER}.$e.sourcecolumns " +
+                  "is not a valid column in table.")
+          }
+        }
+
+        tableProperties.put(CarbonCommonConstants.INDEX_HANDLER +
+          s".$e.sourcecolumntypes", sourceTypes.dropRight(1).toString())
+
+        val handlerType = tableProperties.get(CarbonCommonConstants.INDEX_HANDLER + s".$e.type")
+        val handlerClass = tableProperties.get(CarbonCommonConstants.INDEX_HANDLER + s".$e.class")
+
+        val handlerClassName: String = handlerClass match {
+          case Some(className) =>
+            className
+          case None =>
+            /* use handler type to find the default implementation */
+            if (handlerType.isEmpty) {
+              throw new MalformedCarbonCommandException(
+                s"Carbon ${CarbonCommonConstants.INDEX_HANDLER} property is invalid. " +
+                  s"Both ${CarbonCommonConstants.INDEX_HANDLER}.$e.class and " +
+                  s"${CarbonCommonConstants.INDEX_HANDLER}.$e.type properties are not specified")
+            } else if (handlerType.get.equalsIgnoreCase("geohash")) {
+              /* Use geoHash default implementation */
+              val className = classOf[org.apache.carbondata.core.util.GeoHashDefault].getName
+              tableProperties.put(s"${CarbonCommonConstants.INDEX_HANDLER}.$e.class", className)
+              className
+            } else {
+              throw new MalformedCarbonCommandException(
+                s"Carbon ${CarbonCommonConstants.INDEX_HANDLER} property is invalid. " +
```

Comment:
same as above, better to recheck all the error messages again.

Reply:
Modified